inline migrate fails on security

## Issue description

INLINE migration fails due to:
ERROR sonia.scm.migration.UpdateException - could not execute update for type sonia.scm.security.xml to version 2.0.1 in class sonia.scm.update.security.XmlSecurityV1UpdateStep
java.lang.NullPointerException: null

### Bug Report

2 bugs:
- cannot restart scm-server
- cannot migrate INLINE old 1.60 repositories

Please note:
dpkg -l |grep jre
ii  adoptopenjdk-8-hotspot-jre            8u252-b09-2                          amd64        OpenJDK Runtime Environment 8 (JRE) with Hotspot by AdoptOpenJDK
ii  default-jre                           2:1.11-71                            amd64        Standard Java or Java compatible Runtime
ii  default-jre-headless                  2:1.11-71                            amd64        Standard Java or Java compatible Runtime (headless)
ii  openjdk-11-jre:amd64                  11.0.7+10-3~deb10u1                  amd64        OpenJDK Java runtime, using Hotspot JIT
ii  openjdk-11-jre-headless:amd64         11.0.7+10-3~deb10u1                  amd64        OpenJDK Java runtime, using Hotspot JIT (headless)

But I assume this is not an issue as:
java -version
openjdk version "11.0.7" 2020-04-14
OpenJDK Runtime Environment (build 11.0.7+10-post-Debian-3deb10u1)
OpenJDK 64-Bit Server VM (build 11.0.7+10-post-Debian-3deb10u1, mixed mode, sharing)


#### Describe how to reproduce the bug

mv /var/lib/scm /var/lib/scm2
mkdir /var/lib/scm
copy files/dirs from home-dir of SCM 1.60 to /var/lib/scm  (1.60 SCM server is a different host, SCM 2 server is a fresh install on new server)
chown -R scm:scm /var/lib/scm
browse to scm url
select INLINE migration for all projects
issue on a terminal after some time manually 'service scm-server restart' as _scm-manager GUI cannot restart_ due to:
ERROR sonia.scm.lifecycle.PosixRestartStrategy - failed to collect java vm arguments
java.io.IOException: **Failed to exec 'jsvc.exec' No such file or directory**

the jsvc package is not(!) a **dependency**. But manually installing it does not fix this error.

when checking /var/log/scm/scm-server.out all seems ok after restart until the security.xml needs to be migrated.
End result: non functional scm-manager

**Note**: /var/log/scm/scm-manager.log and /var/log/scm/scm-server.out are (nearly?) identical?

#### Describe your SCM-Manager setup

SCM-manager behind a NGINX server using Let's Encrypt SSL certs

#### Which operating system, browser and versions do you use?

Debian 10 (buster), both on server and client. Browser: chromium from .deb repo: 80.0.3987.162-1~deb10u1

#### Log of the bug

2020-06-28 15:07:12.188 [main] [          ] INFO  sonia.scm.update.UpdateEngine - running update step for type sonia.scm.plugin.svn and version 2.0.0 (class sonia.scm.repository.SvnV2UpdateStep)
2020-06-28 15:07:12.199 [main] [          ] INFO  sonia.scm.update.UpdateEngine - running update step for type sonia.scm.plugin-center and version 2.0.0 (class sonia.scm.update.plugin.PluginCenterUpdateStep)
2020-06-28 15:07:12.215 [main] [          ] INFO  sonia.scm.update.UpdateEngine - running update step for type sonia.scm.user.xml and version 2.0.0 (class sonia.scm.update.user.XmlUserV1UpdateStep)
2020-06-28 15:07:12.215 [main] [          ] INFO  sonia.scm.update.user.XmlUserV1UpdateStep - moved old users file to /var/lib/scm/config/usersV1.xml
2020-06-28 15:07:12.215 [main] [          ] INFO  sonia.scm.update.user.XmlUserV1UpdateStep - moved old security file to /var/lib/scm/config/securityV1.xml
2020-06-28 15:07:12.275 [main] [          ] INFO  sonia.scm.update.UpdateEngine - running update step for type sonia.scm.git.lfs and version 2.0.0 (class sonia.scm.web.lfs.LfsV1UpdateStep)
2020-06-28 15:07:12.278 [main] [          ] INFO  sonia.scm.update.UpdateEngine - running update step for type sonia.scm.group.xml and version 2.0.0 (class sonia.scm.update.group.XmlGroupV1UpdateStep)
2020-06-28 15:07:12.279 [main] [          ] INFO  sonia.scm.update.group.XmlGroupV1UpdateStep - moved old groups file to /var/lib/scm/config/groupsV1.xml
2020-06-28 15:07:12.290 [main] [          ] INFO  sonia.scm.update.UpdateEngine - running update step for type sonia.scm.plugin.git and version 2.0.0 (class sonia.scm.repository.update.GitV2UpdateStep)
2020-06-28 15:07:15.532 [main] [          ] INFO  sonia.scm.update.UpdateEngine - running update step for type sonia.scm.security.xml and version 2.0.1 (class sonia.scm.update.security.XmlSecurityV1UpdateStep)
2020-06-28 15:07:15.556 [main] [          ] ERROR sonia.scm.migration.UpdateException - could not execute update for type sonia.scm.security.xml to version 2.0.1 in class sonia.scm.update.security.XmlSecurityV1UpdateStep
java.lang.NullPointerException: null
	at sonia.scm.update.security.XmlSecurityV1UpdateStep.mapV1Permissions(XmlSecurityV1UpdateStep.java:93)
	at sonia.scm.update.security.XmlSecurityV1UpdateStep.doUpdate(XmlSecurityV1UpdateStep.java:79)
	at sonia.scm.update.UpdateEngine.execute(UpdateEngine.java:87)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)
	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:177)
	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1654)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:497)
	at sonia.scm.update.UpdateEngine.update(UpdateEngine.java:77)
	at sonia.scm.lifecycle.BootstrapContextListener.processUpdates(BootstrapContextListener.java:175)
	at sonia.scm.lifecycle.BootstrapContextListener.startEitherMigrationOrApplication(BootstrapContextListener.java:133)
	at sonia.scm.lifecycle.BootstrapContextListener.createMigrationOrNormalModuleProvider(BootstrapContextListener.java:124)
	at sonia.scm.lifecycle.BootstrapContextListener.createStageTwoInjector(BootstrapContextListener.java:117)
	at sonia.scm.lifecycle.BootstrapContextListener.getInjector(BootstrapContextListener.java:93)
	at com.google.inject.servlet.GuiceServletContextListener.contextInitialized(GuiceServletContextListener.java:45)
	at sonia.scm.lifecycle.BootstrapContextListener.contextInitialized(BootstrapContextListener.java:75)
	at sonia.scm.lifecycle.BootstrapContextFilter.initializeContext(BootstrapContextFilter.java:74)
	at sonia.scm.lifecycle.BootstrapContextFilter.init(BootstrapContextFilter.java:65)
	at org.eclipse.jetty.servlet.FilterHolder.initialize(FilterHolder.java:134)
	at org.eclipse.jetty.servlet.ServletHandler.lambda$initialize$0(ServletHandler.java:749)
	at java.base/java.util.Spliterators$ArraySpliterator.forEachRemaining(Spliterators.java:948)
	at java.base/java.util.stream.Streams$ConcatSpliterator.forEachRemaining(Streams.java:734)
	at java.base/java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:658)
	at org.eclipse.jetty.servlet.ServletHandler.initialize(ServletHandler.java:773)
	at org.eclipse.jetty.servlet.ServletContextHandler.startContext(ServletContextHandler.java:360)
	at org.eclipse.jetty.webapp.WebAppContext.startWebapp(WebAppContext.java:1445)
	at org.eclipse.jetty.webapp.WebAppContext.startContext(WebAppContext.java:1409)
	at org.eclipse.jetty.server.handler.ContextHandler.doStart(ContextHandler.java:844)
	at org.eclipse.jetty.servlet.ServletContextHandler.doStart(ServletContextHandler.java:275)
	at org.eclipse.jetty.webapp.WebAppContext.doStart(WebAppContext.java:524)
	at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:72)
	at org.eclipse.jetty.util.component.ContainerLifeCycle.start(ContainerLifeCycle.java:169)
	at org.eclipse.jetty.util.component.ContainerLifeCycle.doStart(ContainerLifeCycle.java:117)
	at org.eclipse.jetty.server.handler.AbstractHandler.doStart(AbstractHandler.java:97)
	at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:72)
	at org.eclipse.jetty.util.component.ContainerLifeCycle.start(ContainerLifeCycle.java:169)
	at org.eclipse.jetty.server.Server.start(Server.java:408)
	at org.eclipse.jetty.util.component.ContainerLifeCycle.doStart(ContainerLifeCycle.java:110)
	at org.eclipse.jetty.server.handler.AbstractHandler.doStart(AbstractHandler.java:97)
	at org.eclipse.jetty.server.Server.doStart(Server.java:372)
	at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:72)
	at sonia.scm.server.ScmServer.init(ScmServer.java:130)
	at sonia.scm.server.ScmServerDaemon.init(ScmServerDaemon.java:110)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.apache.commons.daemon.support.DaemonLoader.load(DaemonLoader.java:205)
2020-06-28 15:07:15.557 [main] [          ] ERROR sonia.scm.lifecycle.BootstrapContextListener - failed to create stage two injector
sonia.scm.migration.UpdateException: could not execute update for type sonia.scm.security.xml to version 2.0.1 in class sonia.scm.update.security.XmlSecurityV1UpdateStep
	at sonia.scm.update.UpdateEngine.execute(UpdateEngine.java:90)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)
	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:177)
	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1654)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:497)
	at sonia.scm.update.UpdateEngine.update(UpdateEngine.java:77)
	at sonia.scm.lifecycle.BootstrapContextListener.processUpdates(BootstrapContextListener.java:175)
	at sonia.scm.lifecycle.BootstrapContextListener.startEitherMigrationOrApplication(BootstrapContextListener.java:133)
	at sonia.scm.lifecycle.BootstrapContextListener.createMigrationOrNormalModuleProvider(BootstrapContextListener.java:124)
	at sonia.scm.lifecycle.BootstrapContextListener.createStageTwoInjector(BootstrapContextListener.java:117)
	at sonia.scm.lifecycle.BootstrapContextListener.getInjector(BootstrapContextListener.java:93)
	at com.google.inject.servlet.GuiceServletContextListener.contextInitialized(GuiceServletContextListener.java:45)
	at sonia.scm.lifecycle.BootstrapContextListener.contextInitialized(BootstrapContextListener.java:75)
	at sonia.scm.lifecycle.BootstrapContextFilter.initializeContext(BootstrapContextFilter.java:74)
	at sonia.scm.lifecycle.BootstrapContextFilter.init(BootstrapContextFilter.java:65)
	at org.eclipse.jetty.servlet.FilterHolder.initialize(FilterHolder.java:134)
	at org.eclipse.jetty.servlet.ServletHandler.lambda$initialize$0(ServletHandler.java:749)
	at java.base/java.util.Spliterators$ArraySpliterator.forEachRemaining(Spliterators.java:948)
	at java.base/java.util.stream.Streams$ConcatSpliterator.forEachRemaining(Streams.java:734)
	at java.base/java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:658)
	at org.eclipse.jetty.servlet.ServletHandler.initialize(ServletHandler.java:773)
	at org.eclipse.jetty.servlet.ServletContextHandler.startContext(ServletContextHandler.java:360)
	at org.eclipse.jetty.webapp.WebAppContext.startWebapp(WebAppContext.java:1445)
	at org.eclipse.jetty.webapp.WebAppContext.startContext(WebAppContext.java:1409)
	at org.eclipse.jetty.server.handler.ContextHandler.doStart(ContextHandler.java:844)
	at org.eclipse.jetty.servlet.ServletContextHandler.doStart(ServletContextHandler.java:275)
	at org.eclipse.jetty.webapp.WebAppContext.doStart(WebAppContext.java:524)
	at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:72)
	at org.eclipse.jetty.util.component.ContainerLifeCycle.start(ContainerLifeCycle.java:169)
	at org.eclipse.jetty.util.component.ContainerLifeCycle.doStart(ContainerLifeCycle.java:117)
	at org.eclipse.jetty.server.handler.AbstractHandler.doStart(AbstractHandler.java:97)
	at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:72)
	at org.eclipse.jetty.util.component.ContainerLifeCycle.start(ContainerLifeCycle.java:169)
	at org.eclipse.jetty.server.Server.start(Server.java:408)
	at org.eclipse.jetty.util.component.ContainerLifeCycle.doStart(ContainerLifeCycle.java:110)
	at org.eclipse.jetty.server.handler.AbstractHandler.doStart(AbstractHandler.java:97)
	at org.eclipse.jetty.server.Server.doStart(Server.java:372)
	at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:72)
	at sonia.scm.server.ScmServer.init(ScmServer.java:130)
	at sonia.scm.server.ScmServerDaemon.init(ScmServerDaemon.java:110)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.apache.commons.daemon.support.DaemonLoader.load(DaemonLoader.java:205)
Caused by: java.lang.NullPointerException: null
	at sonia.scm.update.security.XmlSecurityV1UpdateStep.mapV1Permissions(XmlSecurityV1UpdateStep.java:93)
	at sonia.scm.update.security.XmlSecurityV1UpdateStep.doUpdate(XmlSecurityV1UpdateStep.java:79)
	at sonia.scm.update.UpdateEngine.execute(UpdateEngine.java:87)
	... 49 common frames omitted

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -14,6 +14,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Fixed
 - Fixed installation of debian packages on distros without preinstalled `at` ([#1216](https://github.com/scm-manager/scm-manager/issues/1216) and [#1217](https://github.com/scm-manager/scm-manager/pull/1217))
+- Fixed broken migration with empty security.xml ([#1219](https://github.com/scm-manager/scm-manager/issues/1219) and [#1221](https://github.com/scm-manager/scm-manager/pull/1221))
 
 ## [2.1.1] - 2020-06-23
 ### Fixed

--- a/scm-webapp/src/main/java/sonia/scm/update/security/XmlSecurityV1UpdateStep.java
+++ b/scm-webapp/src/main/java/sonia/scm/update/security/XmlSecurityV1UpdateStep.java
@@ -21,7 +21,7 @@
  * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
  * SOFTWARE.
  */
-    
+
 package sonia.scm.update.security;
 
 import org.slf4j.Logger;
@@ -44,6 +44,7 @@ import javax.xml.bind.annotation.XmlElement;
 import javax.xml.bind.annotation.XmlRootElement;
 import java.io.File;
 import java.nio.file.Path;
+import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.List;
 import java.util.function.Consumer;
@@ -179,7 +180,7 @@ public class XmlSecurityV1UpdateStep implements UpdateStep {
   @XmlRootElement(name = "configuration")
   private static class V1Security {
     @XmlElement(name = "entry")
-    private List<Entry> entries;
+    private List<Entry> entries = new ArrayList<>();
   }
 
   @XmlAccessorType(XmlAccessType.FIELD)

--- a/scm-webapp/src/test/java/sonia/scm/update/security/XmlSecurityV1UpdateStepTest.java
+++ b/scm-webapp/src/test/java/sonia/scm/update/security/XmlSecurityV1UpdateStepTest.java
@@ -43,6 +43,7 @@ import java.net.URL;
 import java.nio.file.Files;
 import java.nio.file.Path;
 import java.util.List;
+import java.util.Map;
 
 import static java.util.stream.Collectors.toList;
 import static org.assertj.core.api.Assertions.assertThat;
@@ -107,15 +108,17 @@ class XmlSecurityV1UpdateStepTest {
   @Nested
   class WithExistingSecurityXml {
 
+    private Path configDir;
+
     @BeforeEach
     void createSecurityV1XML(@TempDir Path tempDir) throws IOException {
-      Path configDir = tempDir.resolve("config");
+      configDir = tempDir.resolve("config");
       Files.createDirectories(configDir);
-      copyTestDatabaseFile(configDir, "securityV1.xml");
     }
 
     @Test
-    void shouldMapV1PermissionsFromSecurityV1XML() throws JAXBException {
+    void shouldMapV1PermissionsFromSecurityV1XML() throws IOException, JAXBException {
+      copyTestDatabaseFile(configDir, "securityV1.xml");
       updateStep.doUpdate();
       List<String> assignedPermission =
         assignedPermissionStore.getAll().values()
@@ -127,15 +130,27 @@ class XmlSecurityV1UpdateStepTest {
       assertThat(assignedPermission).contains("test");
     }
 
+    @Test
+    void shouldNotFailOnEmptyV1SecurityXml() throws IOException, JAXBException {
+      copyTestDatabaseFile(configDir, "emptySecurityV1.xml", "securityV1.xml");
+      updateStep.doUpdate();
+      assertThat(assignedPermissionStore.getAll()).isEmpty();
+    }
+
   }
 
   private void copyTestDatabaseFile(Path configDir, String fileName) throws IOException {
-    URL url = Resources.getResource("sonia/scm/update/security/" + fileName);
-    Files.copy(url.openStream(), configDir.resolve(fileName));
+    copyTestDatabaseFile(configDir, fileName, fileName);
+  }
+
+  private void copyTestDatabaseFile(Path configDir, String sourceFileName, String targetFileName) throws IOException {
+    URL url = Resources.getResource("sonia/scm/update/security/" + sourceFileName);
+    Files.copy(url.openStream(), configDir.resolve(targetFileName));
   }
 
   @Test
   void shouldNotFailForMissingConfigDir() throws JAXBException {
     updateStep.doUpdate();
+    assertThat(assignedPermissionStore.getAll()).isEmpty();
   }
 }

--- a/scm-webapp/src/test/resources/sonia/scm/update/security/emptySecurityV1.xml
+++ b/scm-webapp/src/test/resources/sonia/scm/update/security/emptySecurityV1.xml
@@ -1,0 +1,28 @@
+<?xml version="1.0" ?>
+<!--
+
+    MIT License
+
+    Copyright (c) 2020-present Cloudogu GmbH and Contributors
+
+    Permission is hereby granted, free of charge, to any person obtaining a copy
+    of this software and associated documentation files (the "Software"), to deal
+    in the Software without restriction, including without limitation the rights
+    to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+    copies of the Software, and to permit persons to whom the Software is
+    furnished to do so, subject to the following conditions:
+
+    The above copyright notice and this permission notice shall be included in all
+    copies or substantial portions of the Software.
+
+    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+    IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+    FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+    AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+    LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+    OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+    SOFTWARE.
+
+-->
+<configuration>
+</configuration>

